vpr/src/place/move_utils.cpp

```diff
@@ -632,9 +637,17 @@ bool find_to_loc_uniform(t_logical_block_type_ptr type,
 
     auto to_type = grid[to.x][to.y].type;
 
-    //Each x/y location contains only a single type, so we can pick a random
-    //z (capcity) location
-    to.z = vtr::irand(to_type->capacity - 1);
+    //Each x/y location possibly contains multiple sub tiles, so we need to pick
+    //a z location within a compatible sub tile.
+    std::vector<int> compatible_sub_tiles;
+    for (int capacity = 0; capacity < to_type->capacity; capacity++) {
+        t_pl_loc loc(to.x, to.y, capacity);
+
+        if (is_tile_compatible(to_type, type, loc)) {
+            compatible_sub_tiles.push_back(capacity);
+        }
+    }
+    to.sub_tile = compatible_sub_tiles[vtr::irand((int)compatible_sub_tiles.size() - 1)];
```

Comment:
This is allocating a a std::vector deep in the placer's inner loop, which is likely to be a significant performance issue. 

Probably we'll need to pull this out and pre-determine what sub-tile locations are compatible with each logical type. I expect this will end-up in the compressed grid data structures.

Reply:
As suggested I have moved the creation of the vector during create_compressed_block_grids.

To do so an unordered map is necessary, as it may be possible that a logical block is can be placed in different tiles with different capacities, hence, the sub tile index depends not only on the logical block, but also on the physical tile chosen for the swap.

This should reduce the impact on run time, and I will run the qor benchmarks to verify this.
If the performance hit is too much, we could just let the move generator produce a random sub tile index which is checked at a later stage out of the inner loop.